### PR TITLE
[ROCm][Perf] Kimi-K3 Fused kernel for KDA decode

### PR DESCRIPTION
**Pull request by kliuae**
## Purpose
Kimi's KDA layers do three things per decode step for every sequence and every value head:

1. Causal conv1d update over the Q/K/V projections, advancing a small per-sequence convolution state
2. Gated delta-rule recurrence, which updates a `[head_dim, head_dim]` recurrent state and produces the attention output
3. gated RMSNorm on that output

On ROCm, currently these run as separate Triton kernels. This PR mirrors the CUDA path, adds a HIP equivalent of the fusion kernel for single-token decode steps. It shows speedup compared with separate Triton kernel launches, and shows improvement to output throughput by ~7% in e2e benchmark.

This kernel has only been tested on gfx950 at the moment, so its build and dispatch is guarded on gfx950 for now.

Kernel latency
| rows | Triton chain | this kernel | speedup |
|---|---|---|---|
| 1 | 10.98 µs | **5.76 µs** | 1.91× |
| 8 | 14.21 | **6.15** | 2.31× |
| 32 | 21.98 | **12.15** | 1.81× |
| 64 | 31.94 | **19.99** | 1.60× |
| 128 | 51.14 | **37.72** | 1.36× |

## Test Plan
lm_eval with gsm8k on MI355
Unit testing for the new kernel

Server command
```
AITER_BF16_FP8_MOE_BOUND=0 \
VLLM_ROCM_USE_AITER=1 \
SAFETENSORS_FAST_GPU=1 \
AITER_SITUV2_A8W4=1 \
VLLM_USE_BREAKABLE_CUDAGRAPH=0 \
vllm serve moonshotai/Kimi-K3 \
  --trust-remote-code \
  --moe-backend auto \
  --tensor-parallel-size 8 \
  --load-format auto \
  --gpu-memory-utilization 0.95 \
  --mm-encoder-tp-mode data \
  --max-num-seqs 128 \
  --max-num-batched-tokens 4096 \
  --enable-auto-tool-choice \
  --tool-call-parser kimi_k3 \
  --reasoning-parser kimi_k3
```

Benchmark
```
vllm bench serve \
  --model moonshotai/Kimi-K3 --backend vllm --trust-remote-code \
  --dataset-name random \
  --random-input-len 8192 --random-output-len 1024 \
  --max-concurrency $CONC --num-prompts $((CONC * 4)) \
  --temperature 0 --ignore-eos --port 8000
```

lm_eval
```
lm_eval --model local-chat-completions \
  --model_args "model=moonshotai/Kimi-K3,base_url=http://localhost:8000/v1/chat/completions,num_concurrent=64,max_retries=3,tokenized_requests=False,max_gen_toks=8192,timeout=3600" \
  --tasks gsm8k --num_fewshot 5 \
  --apply_chat_template --fewshot_as_multiturn \
  --seed 1234
```

### Baseline (with fixes from 50649)

5-shot
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9689|±  |0.0048|
|     |       |strict-match    |     5|exact_match|↑  |0.9697|±  |0.0047|

20-shot
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9697|±  |0.0047|
|     |       |strict-match    |    20|exact_match|↑  |0.9704|±  |0.0047|

### PR

5-shot
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9689|±  |0.0048|
|     |       |strict-match    |     5|exact_match|↑  |0.9697|±  |0.0047|

20-shot
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9689|±  |0.0048|
|     |       |strict-match    |    20|exact_match|↑  |0.9697|±  |0.0047|

E2E performance on MI355X
ISL/OSL | Conc | Config | Total tokens/s | Delta (%) | Total reqs/s | Delta (%) | Output tokens/s | Delta (%) | mean TTFT (ms) | delta (%) | mean TPOT (ms) | delta (%) | mean ITL (ms) | delta (%)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
8192/1024 | 64 | main | 7036.48 | — | 0.7635 | — | 781.83 | — | 6211.72 | — | 75.26 | — | 75.26 | —
8192/1024 | 64 | PR | 7513.44 | +6.78 | 0.8153 | +6.82 | 834.83 | +6.78 | 6185.32 | -0.43 | 70.15 | -6.79 | 70.15 | -6.79

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by Fangzhou-Ai** [comment]
Hi @kliuae , can you please re-benchmark you PR with this updated recipe? We added one more flag to enable automatic fusion .
https://github.com/vllm-project/recipes/pull/733

File: vllm/model_executor/layers/mamba/gdn/kimi_gdn_linear_attn.py [hunk 1]
```diff
@@ -556,10 +640,53 @@ def _prefill_conv(
...
+                        q=q_ns[:, :nd_tok],
+                        k=k_ns[:, :nd_tok],
+                        v=v_ns[:, :nd_tok],
+                        raw_g=g1_ns[:, :nd_tok],
+                        raw_beta=beta_ns[:, :nd_tok],
+                        A_log=self.A_log,
+                        dt_bias=self.dt_bias,
+                        lower_bound=self.gate_lower_bound,
+                        initial_state=recurrent_state,
+                        cu_seqlens=non_spec_query_start_loc[: m.num_decodes + 1],
```
Comment (tjtanaa) [inline vllm/model_executor/layers/mamba/gdn/kimi_gdn_linear_attn.py]: @kliuae please fix the prefix commit

**Comment by Fangzhou-Ai** [comment]
The latest fixed-cohort C16/C24 trace provides a useful estimate of the critical-path budget targeted by this fusion.

Profile setup: 8 x MI355X (gfx950), TP8, BF16/MXFP4, non-speculative decode, prefix caching, 63,911-token shared prefix + 4,089-token suffix, full HIP-graph replay, and `+fused_rms_norm_gated`. The baseline was vLLM `25e1f81e1b8def93d7228a38b83348880babbe7b` with AITER `61549c5c2623316d24ce54dcf35438173cb4d95b`. The fixed-cohort pure-GPU step periods were 32.82 ms at C16 and 35.98 ms at C24.

Across all 69 KDA layers, the exact baseline chain replaced by this PR costs approximately:

| Component | C16 | C24 |
|---|---:|---:|
| causal-conv update | 0.414 ms | 0.425 ms |
| recurrent KDA | 0.603 ms | 0.826 ms |
| recurrent-output placement copy | 0.367 ms | 0.418 ms |
| fused gated RMSNorm | 0.412 ms | 0.424 ms |
| Total targeted budget | 1.80 ms | 2.09 ms |

The total is the baseline targeted-cost envelope and therefore an upper bound on possible saving, not the expected gain, because the fused kernel still performs the recurrence and state traffic. Combining these traced budgets with the PR's nearby-shape microbenchmark speedups (roughly 1.8-2.3x around C16/C24) suggests a reasonable target of:

- C16: about 0.8-1.0 ms/step, or roughly 2.5-3.1% of the pure-GPU period
- C24: about 0.9-1.2 ms/step, or roughly 2.6-3.3% of the pure-GPU period

These are trace-derived expectations rather than measurements of this PR; an exact C16/C24 A/B is still needed. The direct `out=core_attn_out` path also removes the 69 recurrent-output placement copies for gfx950 pure decode. The later KDA output-projection copy is separate and is addressed by #50592, so its gain should not be added to the targeted budget above.

File: vllm/model_executor/layers/mamba/gdn/kimi_gdn_linear_attn.py [hunk 2]
```diff
@@ -250,16 +254,53 @@ def __init__(
...
         self.conv1d.weight.data = self.conv1d.weight.data.unsqueeze(1)
         delattr(self.conv1d.weight, "weight_loader")
-        set_weight_attrs(
-            self.conv1d.weight,
-            {
-                "weight_loader": _make_fused_conv1d_weight_loader(
-                    [self.projection_size] * 3,
-                    self.tp_size,
-                    self.tp_rank,
+        # ROCm can fuse the whole decode step (conv + recurrence + gated norm)
```
Comment (tjtanaa) [inline vllm/model_executor/layers/mamba/gdn/kimi_gdn_linear_attn.py]: can all these logic moved to the `vllm/models/kimi_k3/amd/kda.py` now that this PR https://github.com/vllm-project/vllm/pull/50592 has been merged

File: tests/models/kimi_k3/test_amd_kda_decode.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+conv state and the recurrent state in place. Every one of those outputs is
+compared against the three-kernel path the AMD layer falls back to.
+"""
+
+import pytest
+import torch
+
+from vllm.platforms import current_platform
+
+pytestmark = pytest.mark.skipif(
```
Comment (tjtanaa) [inline tests/models/kimi_k3/test_amd_kda_decode.py:25]: @kliuae we should also skip this test if it is not on gfx950.

but we can only import `on_gfx950` after we guard against `current_platform.is_rocm()`

Comment (kliuae) [reply tests/models/kimi_k3/test_amd_kda_decode.py:25]: Guarded the tests for gfx950

**Review by tjtanaa — APPROVED** [review]
LGTM.
